cpp/src/strings/strings_column_factories.cu

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/strings/strings_column_factories.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/transform_reduce.h>
+#include <thrust/transform_scan.h>
+#include <thrust/for_each.h>
+
+namespace cudf {
+
+// Create a strings-type column.
+// A strings-column has children columns to manage the variable-length
+// encoded character array.
+// Use the strings_column_handler class to perform strings operations
+// on this type of column.
+std::unique_ptr<column> make_strings_column(
+    std::pair<const char*,size_t>* strings, size_type count, cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr)
+{
+    // maybe a separate factory for creating null strings-column
+    CUDF_EXPECTS(count > 0, "must have at least one pair");
+
+    auto execpol = rmm::exec_policy(stream);
+    auto strs = thrust::device_pointer_cast(reinterpret_cast<thrust::pair<const char*,size_t>*>(strings));
+    auto d_strs = strs.get();
+
+    // build offsets column
+    auto offsets_column = make_numeric_column( data_type{INT32}, count, mask_state::UNALLOCATED, stream, mr );
+    auto offsets_view = offsets_column->mutable_view();
+    thrust::transform_inclusive_scan( execpol->on(stream),
+        thrust::make_counting_iterator<size_type>(0), thrust::make_counting_iterator<size_type>(count),
+        offsets_view.data<int32_t>(),
+        [d_strs] __device__ (size_type idx) {
+            thrust::pair<const char*,size_t> item = d_strs[idx];
+            return ( item.first ? (int32_t)item.second : 0 );
+        },
+        thrust::plus<int32_t>() );
+
+    // get number of bytes (last offset value)
+    size_type bytes = thrust::device_pointer_cast(offsets_view.data<int32_t>())[count-1];
+
+    // count nulls
+    size_type null_count = thrust::transform_reduce( execpol->on(stream),
```

Comment:
This all should be able to be replaced with a `valid_if`: https://github.com/rapidsai/cudf/blob/branch-0.10/cpp/src/bitmask/valid_if.cuh#L147